src/core/fs.rs

```diff
@@ -117,3 +121,246 @@ pub fn set_group(path: &Path, group: &str) -> Result<(), io::Error> {
+
+        for fd in &fds {
+            set_fd_nonblocking(*fd).unwrap();
+        }
+
+        let data = Arc::new(FileWatcherData {
+            file: file.to_owned(),
+            read_fd: fds[0],
+            write_fd: fds[1],
+            state: Mutex::new(None),
+        });
+
+        let watcher = {
+            let data = Arc::clone(&data);
+
+            notify::recommended_watcher(move |event: Result<notify::Event, notify::Error>| {
+                let event = match event {
+                    Ok(event) => event,
+                    Err(e) => {
+                        warn!("file watcher error: {:?}", e);
+                        return;
+                    }
+                };
+
+                if !event.paths.into_iter().any(|p| p == data.file) {
+                    // skip unrelated events
+                    return;
+                }
+
+                if let Some(state) = data.state.lock().unwrap().as_mut() {
+                    if !state.changed {
+                        state.changed = true;
+
+                        // non-blocking write to wake up the other side
+
+                        let buf: [u8; 1] = [0; 1];
+
+                        // SAFETY: buf pointer and size are valid
+                        let ret = unsafe {
+                            libc::write(data.write_fd, buf.as_ptr() as *const libc::c_void, 1)
+                        };
+                        assert!(ret == 1 || get_errno() == libc::EAGAIN);
+                    }
+                }
+            })
+            .expect("failed to create file watcher")
+        };
+
+        {
+            let mut state = data.state.lock().unwrap();
+
+            let state = state.insert(FileWatcherState {
+                watcher,
+                changed: false,
+            });
+
+            // watch the dir instead of the file, so we can detect file creates
+            if let Err(e) = state
+                .watcher
+                .watch(dir, notify::RecursiveMode::NonRecursive)
+            {
+                warn!("failed to watch {}: {:?}", dir.display(), e);
+            }
+        }
+
+        Ok(Self { data })
+    }
+
+    pub fn file_changed(&self) -> bool {
+        let mut changed = false;
+
+        if let Some(state) = self.data.state.lock().unwrap().as_mut() {
+            // non-blocking read to clear
+
+            let mut buf = [0u8; 128];
+
+            // SAFETY: buf pointer and size are valid
+            let ret = unsafe {
+                libc::read(
+                    self.data.read_fd,
+                    buf.as_mut_ptr() as *mut libc::c_void,
+                    buf.len(),
+                )
+            };
+            assert!(ret >= 0 || get_errno() == libc::EAGAIN);
+
+            changed = state.changed;
+            state.changed = false;
+        }
+
+        changed
+    }
+}
+
+impl Drop for FileWatcher {
+    fn drop(&mut self) {
+        let mut state = self.data.state.lock().unwrap();
```

Comment:
What happens if we can't get `state` here, does this need to panic or could we try a backoff and drop later? Leaving dangling FDs might put us in a bad state (see Notary 😅)

Reply:
The only time `lock()` fails is if the mutex is poisoned, in which case a panic is warranted. The project aborts on panic (we use `panic = abort` in the cargo settings), so no chance of leaking FDs.